src/util/taproot.rs

```diff
@@ -82,12 +92,753 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+            script: script,
+            ver: ver,
+            merkle_branch: TaprootMerkleBranch(vec![]),
+        }
+    }
+
+    // Compute a leaf hash for the given leaf
+    fn hash(&self) -> sha256::Hash {
+        compute_leaf_hash(&self.script, self.ver)
+    }
+}
+
+/// function to compute leaf hash from components
+pub fn compute_leaf_hash(script: &Script, ver: LeafVersion) -> sha256::Hash {
+    let mut eng = TapLeafHash::engine();
+    ver.as_u8()
+        .consensus_encode(&mut eng)
+        .expect("engines don't err");
+    script
+        .consensus_encode(&mut eng)
+        .expect("engines don't err");
+    sha256::Hash::from_engine(eng)
+}
+
+/// helper function to obtain tap tweak
+pub fn compute_tap_tweak(
+    merkle_root: Option<sha256::Hash>,
+    internal_key: schnorr::PublicKey,
+) -> sha256::Hash {
+    let mut eng = TapTweakHash::engine();
+    // always hash the key
+    eng.input(&internal_key.serialize());
+    if let Some(h) = merkle_root {
+        eng.input(&h);
+    } else {
+        // nothing to hash
+    }
+    sha256::Hash::from_engine(eng)
+}
+
+/// The Merkle proof for inclusion of a tree in a taptree hash
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TaprootMerkleBranch(Vec<sha256::Hash>);
+
+impl TaprootMerkleBranch {
+    /// Obtain a reference to inner
+    pub fn as_inner(&self) -> &[sha256::Hash] {
+        &self.0
+    }
+
+    /// Create a merkle proof from slice
+    pub fn from_slice(sl: &[u8]) -> Result<Self, TaprootError> {
+        if sl.len() % TAPROOT_CONTROL_NODE_SIZE != 0 {
+            Err(TaprootError::InvalidMerkleBranchSize(sl.len()))
+        } else if sl.len() > TAPROOT_CONTROL_NODE_SIZE * TAPROOT_CONTROL_MAX_NODE_COUNT {
+            Err(TaprootError::InvalidMerkleTreeDepth(
+                sl.len() / TAPROOT_CONTROL_NODE_SIZE,
+            ))
+        } else {
+            let mut inner = Vec::with_capacity(sl.len() / TAPROOT_CONTROL_NODE_SIZE);
+            let mut start = 0;
+            while start < sl.len() {
+                let elem = sha256::Hash::from_slice(&sl[start..start + TAPROOT_CONTROL_NODE_SIZE])
+                    .expect("must be multiple for 32");
+                inner.push(elem);
+                start += TAPROOT_CONTROL_NODE_SIZE;
+            }
+            Ok(TaprootMerkleBranch(inner))
+        }
+    }
+
+    /// Serialize self as bytes
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(TAPROOT_CONTROL_NODE_SIZE * self.0.len());
+        for hash in self.0.iter() {
+            buf.extend(hash.as_inner());
+        }
+        buf
+    }
+
+    // Internal function to append elements to proof
+    fn push(&mut self, h: sha256::Hash) -> Result<(), TaprootBuilderError> {
+        self.0.push(h);
+        if self.0.len() > TAPROOT_CONTROL_MAX_NODE_COUNT {
+            return Err(TaprootBuilderError::InvalidMerkleTreeDepth(self.0.len()));
+        } else {
+            Ok(())
+        }
```

Comment:
nit: check for error condition before mutating internal state

```suggestion
        if self.0.len() + 1 > TAPROOT_CONTROL_MAX_NODE_COUNT {
            Err(TaprootBuilderError::InvalidMerkleTreeDepth(self.0.len() + 1))
        } else {
            self.0.push(h);
            Ok(())
        }
```

Reply:
Good catch!

Reply:
Rewriting it to `self.0.len() >= TAPROOT_CONTROL_MAX_NODE_COUNT` would avoid overflow if the invariant that `self.0.len() <= TAPROOT_CONTROL_MAX_NODE_COUNT` is ever accidentally broken.